pkg/cmd/pulumi/stack_migrate.go

```diff
@@ -0,0 +1,152 @@
+	s, err := requireStack(ctx, sourceStack, false, opts, false /*setCurrent*/)
+	if err != nil {
+		return err
+	}
+	currentProjectStack, err := loadProjectStack(s)
+	if err != nil {
+		return err
+	}
+
+	var deployment *apitype.UntypedDeployment
+	// Export the latest version of the checkpoint by default. Otherwise, we require that
+	// the backend/stack implements the ability the export previous checkpoints.
+	if version == "" {
+		deployment, err = s.ExportDeployment(ctx)
+		if err != nil {
+			return err
+		}
+	} else {
+		// Check that the stack and its backend supports the ability to do this.
+		be := s.Backend()
+		specificExpBE, ok := be.(backend.SpecificDeploymentExporter)
+		if !ok {
+			return fmt.Errorf("the current backend (%s) does not provide the ability to export previous deployments",
+				be.Name())
+		}
+
+		deployment, err = specificExpBE.ExportDeploymentForVersion(ctx, s, version)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Build decrypter based on the existing secrets provider
+	var decrypter config.Decrypter
+	currentConfig := currentProjectStack.Config
+
+	if currentConfig.HasSecureValue() {
+		dec, decerr := getStackDecrypter(s)
+		if decerr != nil {
+			return decerr
+		}
+		decrypter = dec
+	} else {
+		decrypter = config.NewPanicCrypter()
+	}
+
+	// Get the new backend
+	targetBackend, err := getBackend(ctx, opts, targetURL)
+	if err != nil {
+		return err
+	}
+
+	// Create the target stack
+	if targetStack == "" {
+		targetStack = s.Ref().String()
+	}
+	targetStackRef, err := targetBackend.ParseStackReference(targetStack)
+	if err != nil {
+		return err
+	}
+	t, err := targetBackend.CreateStack(ctx, targetStackRef, nil)
+	if err != nil {
+		return err
+	}
+
+	err = t.ImportDeployment(ctx, deployment)
+	if err != nil {
+		return err
+	}
+
+	// Create the new secrets provider and set to the currentStack
+	if err := createSecretsManager(ctx, t, secretsProvider, false); err != nil {
+		return err
+	}
+	return migrateOldConfigAndCheckpointToNewSecretsProvider(ctx, t, currentConfig, decrypter)
+}
+
+func newStackMigrateCmd() *cobra.Command {
+	var stackName string
+	var target string
+	var version string
+
+	cmd := &cobra.Command{
+		Use:   "migrate <target backend url> [target secret provider]",
+		Args:  cmdutil.MaximumNArgs(2),
+		Short: "Migrate a stack to a different backend",
+		Long: "Migrate a stack to a different backend.\n" +
+			"\n" +
+			"This will export the current (or selected stack) from the backend currently logged into, and then " +
+			"import that stack into the new target backend. This will also ",
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			ctx := commandContext()
+			opts := display.Options{
+				Color: cmdutil.GetGlobalColorization(),
+			}
+
+			secretProvider := ""
+			if len(args) > 1 {
+				secretProvider = args[1]
+			}
+
+			return migrateStack(
+				ctx, opts,
+				getBackend,
+				stackName, version, target,
+				args[0], secretProvider)
```

Comment:
Maybe check that `args` is not empty and report an error if that was the case before getting first element here? Just in case users don't provide any args for `pulumi stack migrate` and get a panic

Reply:
Ah yes, that's what `cmdutil.RangeArgs(1, 2)` is for. Will fix